File: src/estados-neumatico/estados-neumatico.service.ts
import { ConflictException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { CreateEstadosNeumaticoDto } from './dto/create-estados-neumatico.dto';
import { UpdateEstadosNeumaticoDto } from './dto/update-estados-neumatico.dto';
import { DeleteEstadosNeumaticoResponseDto } from './dto/delete-estados-neumatico-response.dto';
import { EstadosNeumatico } from './entities/estados-neumatico.entity';

@Injectable()
export class EstadosNeumaticoService {
  constructor(
    @InjectRepository(EstadosNeumatico)
    private readonly estadosNeumaticoRepository: Repository<EstadosNeumatico>,
  ) {}

  async create(createEstadosNeumaticoDto: CreateEstadosNeumaticoDto): Promise<EstadosNeumatico> {
    await this.validarEstadoUnico(createEstadosNeumaticoDto.estado);

    const nuevoEstado = this.estadosNeumaticoRepository.create(createEstadosNeumaticoDto);
    return await this.estadosNeumaticoRepository.save(nuevoEstado);
  }

  async findAll(): Promise<EstadosNeumatico[]> {
    return await this.estadosNeumaticoRepository.find({
      order: { id: 'ASC' }
    });
  }

  async findOne(id: number): Promise<EstadosNeumatico> {
    const estado = await this.estadosNeumaticoRepository.findOne({ where: { id } });

    if (!estado) {
      throw new NotFoundException(`Estado con ID ${id} no encontrado`);
    }

    return estado;
  }

  async update(id: number, updateEstadosNeumaticoDto: UpdateEstadosNeumaticoDto): Promise<EstadosNeumatico> {
    const estadoExistente = await this.findOne(id);

    if (updateEstadosNeumaticoDto.estado) {
      await this.validarEstadoUnico(updateEstadosNeumaticoDto.estado, id);
    }

    Object.assign(estadoExistente, updateEstadosNeumaticoDto);
    return await this.estadosNeumaticoRepository.save(estadoExistente);
  }

  async remove(id: number): Promise<DeleteEstadosNeumaticoResponseDto> {
    const estadoExistente = await this.findOne(id);

    // Guardar datos antes de eliminar
    const datosParaResponse = {
      id: estadoExistente.id,
      estado: estadoExistente.estado
    };

    await this.estadosNeumaticoRepository.remove(estadoExistente);

    return new DeleteEstadosNeumaticoResponseDto(
      datosParaResponse.id,
      datosParaResponse.estado
    );
  }

  private async validarEstadoUnico(estado: string, idExcluir?: number): Promise<void> {
    const existente = await this.estadosNeumaticoRepository.findOne({
      where: { estado },
    });

    if (existente && existente.id !== idExcluir) {
      throw new ConflictException(`El estado "${estado}" ya existe`);
    }
  }
}

File: src/main/java/com/tallermecanico/service/EstadosNeumaticoService.java
package com.tallermecanico.service;

import com.tallermecanico.dto.request.CreateEstadosNeumaticoDto;
import com.tallermecanico.dto.request.UpdateEstadosNeumaticoDto;
import com.tallermecanico.dto.response.DeleteEstadosNeumaticoResponseDto;
import com.tallermecanico.entity.EstadosNeumatico;
import com.tallermecanico.exception.DuplicateResourceException;
import com.tallermecanico.exception.ResourceNotFoundException;
import com.tallermecanico.repository.EstadosNeumaticoRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class EstadosNeumaticoService {

    private final EstadosNeumaticoRepository estadosNeumaticoRepository;

    public EstadosNeumaticoService(EstadosNeumaticoRepository estadosNeumaticoRepository) {
        this.estadosNeumaticoRepository = estadosNeumaticoRepository;
    }

    @Transactional
    public EstadosNeumatico create(CreateEstadosNeumaticoDto dto) {
        // Validar estado único
        if (estadosNeumaticoRepository.existsByEstado(dto.getEstado())) {
            throw new DuplicateResourceException("EstadosNeumatico", "estado", dto.getEstado());
        }

        EstadosNeumatico estado = new EstadosNeumatico();
        estado.setEstado(dto.getEstado());

        return estadosNeumaticoRepository.save(estado);
    }

    @Transactional(readOnly = true)
    public List<EstadosNeumatico> findAll() {
        return estadosNeumaticoRepository.findAll();
    }

    @Transactional(readOnly = true)
    public EstadosNeumatico findById(Integer id) {
        return estadosNeumaticoRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("EstadosNeumatico", "id", id));
    }

    @Transactional
    public EstadosNeumatico update(Integer id, UpdateEstadosNeumaticoDto dto) {
        EstadosNeumatico estado = findById(id);

        if (dto.getEstado() != null) {
            // Validar estado único (excluyendo el actual)
            if (estadosNeumaticoRepository.existsByEstadoAndIdNot(dto.getEstado(), id)) {
                throw new DuplicateResourceException("EstadosNeumatico", "estado", dto.getEstado());
            }
            estado.setEstado(dto.getEstado());
        }

        return estadosNeumaticoRepository.save(estado);
    }

    @Transactional
    public DeleteEstadosNeumaticoResponseDto delete(Integer id) {
        EstadosNeumatico estado = findById(id);
        estadosNeumaticoRepository.delete(estado);
        return new DeleteEstadosNeumaticoResponseDto(estado.getId(), estado.getEstado());
    }
}
